device/core/src/device_client.ts

```diff
@@ -171,13 +176,85 @@ export class Client extends InternalClient {
       retryOp.retry((opCallback) => {
         /*Codes_SRS_NODE_DEVICE_CLIENT_16_040: [The `uploadToBlob` method shall call the `_callback` callback with an `Error` object if the upload fails.]*/
         /*Codes_SRS_NODE_DEVICE_CLIENT_16_041: [The `uploadToBlob` method shall call the `_callback` callback no parameters if the upload succeeds.]*/
-        this.blobUploadClient.uploadToBlob(blobName, stream, streamLength, opCallback);
+        this._blobUploadClient.uploadToBlob(blobName, stream, streamLength, opCallback);
       }, (err, result) => {
         safeCallback(_callback, err, result);
       });
     }, callback);
   }
 
+  /**
+   * @description      The `getBlobSharedAccessSignature` gets the linked storage account SAS Token from IoT Hub
+   *
+   * @param {String}    blobName                The name to use for the blob that will be created with the content of the stream.
+   * @param {Callback}  [callback]              Optional callback to call when the upload is complete.
+   * @returns {Promise<UploadParams> | void}    Promise if no callback function was passed, void otherwise.
+   *
+   * @throws {ReferenceException} If blobName is falsy.
+   */
+  getBlobSharedAccessSignature(blobName: string, callback: Callback<UploadParams>): void;
+  getBlobSharedAccessSignature(blobName: string): Promise<UploadParams>;
+  getBlobSharedAccessSignature(blobName: string, callback?: Callback<UploadParams>): Promise<UploadParams> | void {
+    return callbackToPromise((_callback) => {
+      /*Codes_SRS_NODE_DEVICE_CLIENT_41_001: [The `getBlobSharedAccessSignature` method shall throw a `ReferenceError` if `blobName` is falsy.]*/
+      if (!blobName) throw new ReferenceError('blobName cannot be \'' + blobName + '\'');
+      const retryOp = new RetryOperation(this._retryPolicy, this._maxOperationTimeout);
+      retryOp.retry((opCallback) => {
+        /*Codes_SRS_NODE_DEVICE_CLIENT_41_002: [The `getBlobSharedAccessSignature` method shall call the `getBlobSharedAccessSignature` method in the instantiated `_fileUploadApi` class and pass in `blobName` as a parameter.]*/
+        this._fileUploadApi.getBlobSharedAccessSignature(blobName, opCallback);
+      }, (err, result) => {
+        /*Codes_SRS_NODE_DEVICE_CLIENT_41_003: [The `getBlobSharedAccessSignature` method shall call the `_callback` callback with `err` and `result` from the call to `getBlobSharedAccessSignature`.]*/
+        if (!err) {
+          debug('got blob storage shared access signature.');
+          /*Codes_SRS_NODE_DEVICE_CLIENT_41_004: [The `getBlobSharedAccessSignature` method shall store the `result` value for use by the `notifyBlobUploadStatus` method.]*/
+          this._blobStorageUploadParams = result;
```

Comment:
Actually I think there's another bigger issue: Since it's overwritten, this won't work if the app calls `getBlobSharedAccessSignature` multiple times.
I think in that case it'll be able to notify only the very last one that won the overwrite.